File: tools/typescript/src/shared/api-client.ts
/**
 * Telnyx API client using native fetch (Node 18+).
 */

export class TelnyxAPIError extends Error {
  readonly statusCode: number;
  readonly detail: string;
  readonly errors: Record<string, unknown>[];
  readonly retryAfterSeconds?: number;

  constructor(
    statusCode: number,
    detail: string,
    errors: Record<string, unknown>[] = [],
    retryAfterSeconds?: number,
  ) {
    super(`Telnyx API error ${statusCode}: ${detail}`);
    this.name = "TelnyxAPIError";
    this.statusCode = statusCode;
    this.detail = detail;
    this.errors = errors;
    this.retryAfterSeconds = retryAfterSeconds;
  }
}

export interface TelnyxAPIClientOptions {
  baseUrl?: string;
  timeout?: number;
}

export interface TelnyxRequestOptions {
  headers?: Record<string, string>;
  idempotencyKey?: string;
}

export interface TelnyxPollOptions {
  params?: Record<string, unknown>;
  headers?: Record<string, string>;
  intervalMs?: number;
  timeoutMs?: number;
  isDone?: (response: Record<string, unknown>) => boolean;
}

type ResponseMeta = {
  body: Record<string, unknown>;
  statusCode: number;
  retryAfterMs?: number;
};

export class TelnyxAPIClient {
  readonly apiKey: string;
  private readonly baseUrl: string;
  private readonly timeout: number;

  constructor(
    apiKey: string,
    options: TelnyxAPIClientOptions = {},
  ) {
    this.apiKey = apiKey;
    this.baseUrl = (options.baseUrl ?? "https://api.telnyx.com/v2").replace(
      /\/$/,
      "",
    );
    this.timeout = options.timeout ?? 30000;
  }

  private getHeaders(options: TelnyxRequestOptions = {}): Record<string, string> {
    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.apiKey}`,
      "Content-Type": "application/json",
      Accept: "application/json",
    };

    if (options.idempotencyKey) {
      headers["Idempotency-Key"] = options.idempotencyKey;
    }

    return {
      ...headers,
      ...options.headers,
    };
  }

  private async handleResponse(
    response: Response,
  ): Promise<ResponseMeta> {
    const retryAfterSeconds = parseRetryAfterSeconds(response.headers.get("retry-after"));
    const retryAfterMs = retryAfterSeconds === undefined ? undefined : retryAfterSeconds * 1000;

    if (response.status >= 400) {
      let detail: string;
      let errors: Record<string, unknown>[] = [];
      try {
        const body = (await response.json()) as Record<string, unknown>;
        const bodyErrors = body.errors as Record<string, unknown>[] | undefined;
        if (bodyErrors && bodyErrors.length > 0) {
          errors = bodyErrors;
          detail = (bodyErrors[0].detail as string) ?? response.statusText;
        } else {
          detail = response.statusText;
        }
      } catch {
        detail = response.statusText;
      }
      throw new TelnyxAPIError(response.status, detail, errors, retryAfterSeconds);
    }

    if (response.status === 204) {
      return { body: {}, statusCode: response.status, retryAfterMs };
    }

    return {
      body: (await response.json()) as Record<string, unknown>,
      statusCode: response.status,
      retryAfterMs,
    };
  }

  private async request(
    method: "GET" | "POST" | "PUT" | "PATCH" | "DELETE",
    path: string,
    options: {
      params?: Record<string, unknown>;
      json?: Record<string, unknown>;
      request?: TelnyxRequestOptions;
    } = {},
  ): Promise<ResponseMeta> {
    let url = `${this.baseUrl}${path}`;

    if (options.params && Object.keys(options.params).length > 0) {
      const searchParams = new URLSearchParams();
      for (const [key, value] of Object.entries(options.params)) {
        if (value === undefined || value === null) continue;
        if (Array.isArray(value)) {
          for (const item of value) {
            searchParams.append(key, String(item));
          }
        } else {
          searchParams.append(key, String(value));
        }
      }
      url += `?${searchParams.toString()}`;
    }

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(url, {
        method,
        headers: this.getHeaders(options.request),
        body: options.json ? JSON.stringify(options.json) : undefined,
        signal: controller.signal,
      });
      return await this.handleResponse(response);
    } finally {
      clearTimeout(timeoutId);
    }
  }

  async get(
    path: string,
    params?: Record<string, unknown>,
  ): Promise<Record<string, unknown>> {
    const response = await this.request("GET", path, { params });
    return response.body;
  }

  async post(
    path: string,
    json?: Record<string, unknown>,
    options: TelnyxRequestOptions = {},
  ): Promise<Record<string, unknown>> {
    const response = await this.request("POST", path, { json, request: options });
    return response.body;
  }

  async put(
    path: string,
    json?: Record<string, unknown>,
    options: TelnyxRequestOptions = {},
  ): Promise<Record<string, unknown>> {
    const response = await this.request("PUT", path, { json, request: options });
    return response.body;
  }

  async patch(
    path: string,
    json?: Record<string, unknown>,
    options: TelnyxRequestOptions = {},
  ): Promise<Record<string, unknown>> {
    const response = await this.request("PATCH", path, { json, request: options });
    return response.body;
  }

  async delete(path: string, options: TelnyxRequestOptions = {}): Promise<Record<string, unknown>> {
    const response = await this.request("DELETE", path, { request: options });
    return response.body;
  }

  async poll(
    path: string,
    options: TelnyxPollOptions = {},
  ): Promise<Record<string, unknown>> {
    const isDone = options.isDone ?? defaultIsDone;
    const intervalMs = options.intervalMs ?? 1000;
    const timeoutMs = options.timeoutMs ?? this.timeout;
    const startedAt = Date.now();

    while (true) {
      const response = await this.request("GET", path, {
        params: options.params,
        request: { headers: options.headers },
      });

      if (isDone(response.body)) {
        return response.body;
      }

      const elapsedMs = Date.now() - startedAt;
      if (elapsedMs >= timeoutMs) {
        throw new Error(`Timed out polling ${path} after ${timeoutMs}ms`);
      }

      const delayMs = response.retryAfterMs ?? intervalMs;
      await sleep(Math.min(delayMs, timeoutMs - elapsedMs));
    }
  }
}

function defaultIsDone(response: Record<string, unknown>): boolean {
  const data = asRecord(response.data);
  const status = firstString(data?.status, response.status, data?.operation_status);
  if (!status) return true;

  return ["completed", "failed", "canceled", "cancelled", "error", "succeeded"].includes(
    status.toLowerCase(),
  );
}

function parseRetryAfterSeconds(value: string | null): number | undefined {
  if (!value) return undefined;

  const numeric = Number(value);
  if (Number.isFinite(numeric) && numeric >= 0) {
    return numeric;
  }

  const retryAt = Date.parse(value);
  if (Number.isNaN(retryAt)) {
    return undefined;
  }

  return Math.max(0, Math.ceil((retryAt - Date.now()) / 1000));
}

function asRecord(value: unknown): Record<string, unknown> | undefined {
  return value !== null && typeof value === "object" ? (value as Record<string, unknown>) : undefined;
}

function firstString(...values: unknown[]): string | undefined {
  for (const value of values) {
    if (typeof value === "string" && value.length > 0) {
      return value;
    }
  }
  return undefined;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, Math.max(0, ms)));
}

File: tools/python/telnyx_agent_toolkit/shared/api_client.py
"""Telnyx API client using httpx."""

from __future__ import annotations

import asyncio
from typing import Any, Callable

import httpx


class TelnyxAPIError(Exception):
    """Error from the Telnyx API."""

    def __init__(
        self,
        status_code: int,
        detail: str,
        errors: list[dict[str, Any]] | None = None,
        retry_after_seconds: float | None = None,
    ) -> None:
        self.status_code = status_code
        self.detail = detail
        self.errors = errors or []
        self.retry_after_seconds = retry_after_seconds
        super().__init__(f"Telnyx API error {status_code}: {detail}")


class TelnyxAPIClient:
    """Async-first HTTP client for the Telnyx API v2.

    Uses httpx.AsyncClient under the hood. Provides both async and sync methods.
    """

    def __init__(
        self,
        api_key: str,
        *,
        base_url: str = "https://api.telnyx.com/v2",
        timeout: float = 30.0,
    ) -> None:
        self._api_key = api_key
        self._base_url = base_url.rstrip("/")
        self._timeout = timeout
        self._async_client: httpx.AsyncClient | None = None

    @property
    def api_key(self) -> str:
        return self._api_key

    def _get_headers(self) -> dict[str, str]:
        return {
            "Authorization": f"Bearer {self._api_key}",
            "Content-Type": "application/json",
            "Accept": "application/json",
        }

    def _build_headers(
        self,
        *,
        headers: dict[str, str] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, str]:
        built = self._get_headers()
        if idempotency_key:
            built["Idempotency-Key"] = idempotency_key
        if headers:
            built.update(headers)
        return built

    async def _get_async_client(self) -> httpx.AsyncClient:
        if self._async_client is None or self._async_client.is_closed:
            self._async_client = httpx.AsyncClient(
                base_url=self._base_url,
                headers=self._get_headers(),
                timeout=self._timeout,
            )
        return self._async_client

    async def _handle_response(self, response: httpx.Response) -> tuple[dict[str, Any], float | None]:
        retry_after_seconds = _parse_retry_after_seconds(response.headers.get("retry-after"))
        if response.status_code >= 400:
            try:
                body = response.json()
                errors = body.get("errors", [])
                detail = errors[0].get("detail", response.text) if errors else response.text
            except Exception:
                detail = response.text
                errors = []
            raise TelnyxAPIError(
                status_code=response.status_code,
                detail=detail,
                errors=errors,
                retry_after_seconds=retry_after_seconds,
            )
        if response.status_code == 204:
            return {}, retry_after_seconds
        return response.json(), retry_after_seconds  # type: ignore[no-any-return]

    async def get_async(
        self,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Async GET request."""
        client = await self._get_async_client()
        response = await client.get(path, params=params, headers=self._build_headers(headers=headers))
        body, _ = await self._handle_response(response)
        return body

    async def post_async(
        self,
        path: str,
        *,
        json: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        """Async POST request."""
        client = await self._get_async_client()
        response = await client.post(
            path,
            json=json,
            headers=self._build_headers(headers=headers, idempotency_key=idempotency_key),
        )
        body, _ = await self._handle_response(response)
        return body

    async def delete_async(
        self,
        path: str,
        *,
        headers: dict[str, str] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        """Async DELETE request."""
        client = await self._get_async_client()
        response = await client.delete(
            path,
            headers=self._build_headers(headers=headers, idempotency_key=idempotency_key),
        )
        body, _ = await self._handle_response(response)
        return body

    async def put_async(
        self,
        path: str,
        *,
        json: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        """Async PUT request."""
        client = await self._get_async_client()
        response = await client.put(
            path,
            json=json,
            headers=self._build_headers(headers=headers, idempotency_key=idempotency_key),
        )
        body, _ = await self._handle_response(response)
        return body

    async def patch_async(
        self,
        path: str,
        *,
        json: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        """Async PATCH request."""
        client = await self._get_async_client()
        response = await client.patch(
            path,
            json=json,
            headers=self._build_headers(headers=headers, idempotency_key=idempotency_key),
        )
        body, _ = await self._handle_response(response)
        return body

    async def poll_async(
        self,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
        interval_seconds: float = 1.0,
        timeout_seconds: float | None = None,
        is_done: Callable[[dict[str, Any]], bool] | None = None,
    ) -> dict[str, Any]:
        """Poll a GET endpoint until it reaches a terminal state."""
        deadline = asyncio.get_running_loop().time() + (timeout_seconds or self._timeout)
        poll_done = is_done or _default_is_done

        while True:
            client = await self._get_async_client()
            response = await client.get(path, params=params, headers=self._build_headers(headers=headers))
            body, retry_after_seconds = await self._handle_response(response)

            if poll_done(body):
                return body

            remaining = deadline - asyncio.get_running_loop().time()
            if remaining <= 0:
                raise TimeoutError(f"Timed out polling {path} after {timeout_seconds or self._timeout} seconds")

            delay = min(retry_after_seconds or interval_seconds, remaining)
            await asyncio.sleep(max(0.0, delay))

    def get(
        self,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Sync GET request (convenience wrapper)."""
        return _run_sync(self.get_async(path, params=params, headers=headers))

    def post(
        self,
        path: str,
        *,
        json: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        """Sync POST request (convenience wrapper)."""
        return _run_sync(self.post_async(path, json=json, headers=headers, idempotency_key=idempotency_key))

    def delete(
        self,
        path: str,
        *,
        headers: dict[str, str] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        """Sync DELETE request (convenience wrapper)."""
        return _run_sync(self.delete_async(path, headers=headers, idempotency_key=idempotency_key))

    def put(
        self,
        path: str,
        *,
        json: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        """Sync PUT request (convenience wrapper)."""
        return _run_sync(self.put_async(path, json=json, headers=headers, idempotency_key=idempotency_key))

    def patch(
        self,
        path: str,
        *,
        json: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        """Sync PATCH request (convenience wrapper)."""
        return _run_sync(self.patch_async(path, json=json, headers=headers, idempotency_key=idempotency_key))

    def poll(
        self,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
        interval_seconds: float = 1.0,
        timeout_seconds: float | None = None,
        is_done: Callable[[dict[str, Any]], bool] | None = None,
    ) -> dict[str, Any]:
        """Sync polling convenience wrapper."""
        return _run_sync(
            self.poll_async(
                path,
                params=params,
                headers=headers,
                interval_seconds=interval_seconds,
                timeout_seconds=timeout_seconds,
                is_done=is_done,
            )
        )

    async def close(self) -> None:
        """Close the async client."""
        if self._async_client and not self._async_client.is_closed:
            await self._async_client.aclose()


def _run_sync(coro: Any) -> Any:
    """Run an async coroutine synchronously.

    Uses a persistent event loop in a background thread to avoid
    the 'Event loop is closed' issue from repeated asyncio.run() calls.
    """
    try:
        loop = asyncio.get_running_loop()
    except RuntimeError:
        loop = None

    if loop and loop.is_running():
        # Already in an async context — create a new thread
        import concurrent.futures

        with concurrent.futures.ThreadPoolExecutor(max_workers=1) as pool:
            return pool.submit(asyncio.run, coro).result()
    else:
        loop = _get_sync_loop()
        return asyncio.run_coroutine_threadsafe(coro, loop).result()


_sync_loop: asyncio.AbstractEventLoop | None = None
_sync_thread: Any = None


def _get_sync_loop() -> asyncio.AbstractEventLoop:
    """Get or create a persistent background event loop for sync calls."""
    global _sync_loop, _sync_thread
    if _sync_loop is None or _sync_loop.is_closed():
        import threading

        _sync_loop = asyncio.new_event_loop()

        def _run_loop(loop: asyncio.AbstractEventLoop) -> None:
            asyncio.set_event_loop(loop)
            loop.run_forever()

        _sync_thread = threading.Thread(target=_run_loop, args=(_sync_loop,), daemon=True)
        _sync_thread.start()
    return _sync_loop


def _parse_retry_after_seconds(value: str | None) -> float | None:
    if not value:
        return None

    try:
        parsed = float(value)
    except ValueError:
        return None

    return parsed if parsed >= 0 else None


def _default_is_done(response: dict[str, Any]) -> bool:
    data = response.get("data")
    if isinstance(data, dict):
        status = data.get("status") or data.get("operation_status")
    else:
        status = response.get("status")

    if not isinstance(status, str) or not status:
        return True

    return status.lower() in {"completed", "failed", "canceled", "cancelled", "error", "succeeded"}
